File: provable-proof-js/src/verify.ts
import {
  verify,
  verifyWithInclusion,
  type VerifyRequest,
  type VerifyWithInclusionRequest,
} from 'provable-sdk-js';
import { KayrosEnvelope } from './envelope';
import type {
  EnvelopeVerifyDetails,
  EnvelopeVerifyOverrides,
  EnvelopeVerifyResult,
  EnvelopeVerifyWithInclusionOverrides,
} from './types';

interface EnvelopeVerifyInput {
  request: VerifyRequest;
  details: Partial<EnvelopeVerifyDetails>;
}

function firstString(...values: unknown[]): string | undefined {
  for (const value of values) {
    if (typeof value === 'string' && value.length > 0) {
      return value;
    }
  }
  return undefined;
}

function normalizeHexString(value: string | undefined): string | undefined {
  if (!value) {
    return undefined;
  }
  const normalized = value.trim().toLowerCase().replace(/^0x/, '');
  if (!normalized || normalized.length % 2 !== 0 || !/^[0-9a-f]+$/.test(normalized)) {
    return undefined;
  }
  return normalized;
}

function invalidEnvelopeResult(
  details: Partial<EnvelopeVerifyDetails>,
  error: string,
): EnvelopeVerifyResult {
  return {
    valid: false,
    error,
    details: {
      lookupMode: details.envelopeKayrosHash ? 'kayros_hash' : 'data_item',
      recordFound: false,
      ...details,
    },
  };
}

export async function buildEnvelopeVerifyRequest(
  envelope: KayrosEnvelope,
  overrides: EnvelopeVerifyOverrides = {},
): Promise<EnvelopeVerifyInput> {
  const computedDataItem = normalizeHexString(await envelope.computeDataHash());
  const envelopeDataItem = normalizeHexString(envelope.getDataHash());
  const envelopeDataType = firstString(
    envelope.getDataType(),
    envelope.getDataTypeLabel(),
  );
  const envelopeKayrosHash = normalizeHexString(envelope.getKayrosHash());

  const request: VerifyRequest = {
    data_type: firstString(overrides.data_type, overrides.dataType, envelopeDataType),
    data_item: normalizeHexString(firstString(overrides.data_item, overrides.dataItem, computedDataItem)),
    kayros_hash: normalizeHexString(firstString(overrides.kayros_hash, overrides.kayrosHash, envelopeKayrosHash)),
    apiKey: firstString(overrides.apiKey, overrides.api_key, overrides.userKey),
  };

  return {
    request,
    details: {
      computedDataItem,
      envelopeDataItem,
      envelopeDataItemMatch: envelopeDataItem && computedDataItem ? envelopeDataItem === computedDataItem : undefined,
      envelopeDataType,
      envelopeKayrosHash,
    },
  };
}

export async function verifyEnvelope(
  envelope: KayrosEnvelope,
  overrides: EnvelopeVerifyOverrides = {},
): Promise<EnvelopeVerifyResult> {
  const input = await buildEnvelopeVerifyRequest(envelope, overrides);
  if (input.details.envelopeDataItem && input.details.envelopeDataItemMatch === false) {
    return invalidEnvelopeResult(
      input.details,
      `Envelope data_item mismatch: expected=${input.details.envelopeDataItem} computed=${input.details.computedDataItem}`,
    );
  }

  const result = await verify(input.request);
  return {
    valid: result.valid,
    error: result.error,
    details: result.details ? { ...result.details, ...input.details } : undefined,
  };
}

export async function verifyEnvelopeWithInclusion(
  envelope: KayrosEnvelope,
  overrides: EnvelopeVerifyWithInclusionOverrides = {},
): Promise<EnvelopeVerifyResult> {
  const input = await buildEnvelopeVerifyRequest(envelope, overrides);
  if (input.details.envelopeDataItem && input.details.envelopeDataItemMatch === false) {
    return invalidEnvelopeResult(
      input.details,
      `Envelope data_item mismatch: expected=${input.details.envelopeDataItem} computed=${input.details.computedDataItem}`,
    );
  }

  const request: VerifyWithInclusionRequest = {
    ...input.request,
    trusted_root_hash: firstString(overrides.trusted_root_hash, overrides.trustedRootHash),
    trusted_level: overrides.trusted_level ?? overrides.trustedLevel,
    trusted_position: overrides.trusted_position ?? overrides.trustedPosition,
    verify_batch_existence: overrides.verify_batch_existence ?? overrides.verifyBatchExistence,
    level_checks: overrides.level_checks ?? overrides.levelChecks,
  };

  const result = await verifyWithInclusion(request);
  return {
    valid: result.valid,
    error: result.error,
    details: result.details ? { ...result.details, ...input.details } : undefined,
  };
}

File: provable-proof-rs/src/verify.rs
use crate::envelope::KayrosEnvelope;
use crate::types::{
    EnvelopeVerifyDetails, EnvelopeVerifyInput, EnvelopeVerifyOverrides, EnvelopeVerifyResult,
    EnvelopeVerifyWithInclusionOverrides,
};
use provable_sdk::{
    verify, verify_with_inclusion, VerifyRequest, VerifyResult, VerifyResultDetails,
    VerifyWithInclusionRequest,
};

fn normalize_hex(value: Option<&str>) -> Option<String> {
    value.and_then(|input| {
        let trimmed = input.trim();
        if trimmed.is_empty() {
            return None;
        }
        let normalized = trimmed.trim_start_matches("0x").to_lowercase();
        if normalized.len() % 2 == 0 && normalized.chars().all(|ch| ch.is_ascii_hexdigit()) {
            Some(normalized)
        } else {
            None
        }
    })
}

fn invalid_envelope_result(details: EnvelopeVerifyDetails, error: String) -> EnvelopeVerifyResult {
    EnvelopeVerifyResult {
        valid: false,
        error: Some(error),
        details: Some(details),
    }
}

fn merge_details(
    base: Option<VerifyResultDetails>,
    extra: EnvelopeVerifyDetails,
) -> EnvelopeVerifyDetails {
    EnvelopeVerifyDetails {
        verify: base.unwrap_or_else(|| {
            let mut details = VerifyResultDetails::default();
            details.lookup_mode = if extra.envelope_kayros_hash.is_some() {
                "kayros_hash".to_string()
            } else {
                "data_item".to_string()
            };
            details.record_found = false;
            details
        }),
        computed_data_item: extra.computed_data_item,
        envelope_data_item: extra.envelope_data_item,
        envelope_data_item_match: extra.envelope_data_item_match,
        envelope_data_type: extra.envelope_data_type,
        envelope_kayros_hash: extra.envelope_kayros_hash,
    }
}

pub fn build_envelope_verify_request(
    envelope: &KayrosEnvelope,
    overrides: &EnvelopeVerifyOverrides,
) -> provable_sdk::Result<EnvelopeVerifyInput> {
    let computed_data_item = normalize_hex(Some(&envelope.compute_data_hash()?));
    let envelope_data_item = envelope.get_data_hash();
    let envelope_data_type = overrides
        .data_type
        .clone()
        .or_else(|| envelope.get_data_type())
        .or_else(|| envelope.get_data_type_label());
    let envelope_kayros_hash = envelope.get_kayros_hash();

    Ok(EnvelopeVerifyInput {
        request: VerifyRequest {
            data_type: envelope_data_type.clone(),
            data_item: overrides
                .data_item
                .clone()
                .or_else(|| computed_data_item.clone())
                .and_then(|value| normalize_hex(Some(&value))),
            kayros_hash: overrides
                .kayros_hash
                .clone()
                .or_else(|| envelope_kayros_hash.clone())
                .and_then(|value| normalize_hex(Some(&value))),
            api_key: overrides.api_key.clone(),
        },
        details: EnvelopeVerifyDetails {
            verify: VerifyResultDetails::default(),
            computed_data_item: computed_data_item.clone(),
            envelope_data_item: envelope_data_item.clone(),
            envelope_data_item_match: match (
                envelope_data_item.as_ref(),
                computed_data_item.as_ref(),
            ) {
                (Some(expected), Some(computed)) => Some(expected == computed),
                _ => None,
            },
            envelope_data_type,
            envelope_kayros_hash,
        },
    })
}

pub fn verify_envelope(
    envelope: &KayrosEnvelope,
    overrides: &EnvelopeVerifyOverrides,
) -> EnvelopeVerifyResult {
    let input = match build_envelope_verify_request(envelope, overrides) {
        Ok(input) => input,
        Err(error) => {
            return EnvelopeVerifyResult {
                valid: false,
                error: Some(error.0),
                details: None,
            }
        }
    };
    if input.details.envelope_data_item_match == Some(false) {
        let expected = input.details.envelope_data_item.clone().unwrap_or_default();
        let computed = input.details.computed_data_item.clone().unwrap_or_default();
        return invalid_envelope_result(
            input.details,
            format!(
                "Envelope data_item mismatch: expected={} computed={}",
                expected, computed
            ),
        );
    }

    let result: VerifyResult = verify(input.request.clone());
    EnvelopeVerifyResult {
        valid: result.valid,
        error: result.error.clone(),
        details: Some(merge_details(result.details, input.details)),
    }
}

pub fn verify_envelope_with_inclusion(
    envelope: &KayrosEnvelope,
    overrides: &EnvelopeVerifyWithInclusionOverrides,
) -> EnvelopeVerifyResult {
    let input = match build_envelope_verify_request(envelope, &overrides.overrides) {
        Ok(input) => input,
        Err(error) => {
            return EnvelopeVerifyResult {
                valid: false,
                error: Some(error.0),
                details: None,
            }
        }
    };
    if input.details.envelope_data_item_match == Some(false) {
        let expected = input.details.envelope_data_item.clone().unwrap_or_default();
        let computed = input.details.computed_data_item.clone().unwrap_or_default();
        return invalid_envelope_result(
            input.details,
            format!(
                "Envelope data_item mismatch: expected={} computed={}",
                expected, computed
            ),
        );
    }
    let request = VerifyWithInclusionRequest {
        verify_request: input.request.clone(),
        trusted_root_hash: overrides.trusted_root_hash.clone(),
        trusted_level: overrides.trusted_level,
        trusted_position: overrides.trusted_position,
        levels_hash_type: overrides.levels_hash_type.clone(),
        verify_batch_existence: overrides.verify_batch_existence,
        level_checks: overrides.level_checks.clone(),
    };
    let result = verify_with_inclusion(request);
    EnvelopeVerifyResult {
        valid: result.valid,
        error: result.error.clone(),
        details: Some(merge_details(result.details, input.details)),
    }
}
